scanner/indexer/fetcher.go

```diff
@@ -0,0 +1,313 @@
+package indexer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	gosync "sync"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/indexer"
+	"github.com/quay/zlog"
+	"github.com/stackrox/scanner/v4/internal/sync"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	_ indexer.FetchArena = (*localFetchArena)(nil)
+	_ indexer.Realizer   = (*localFetcher)(nil)
+)
+
+// localFetchArena implements indexer.FetchArena.
+//
+// It is designed to minimize layer downloads and attempt
+// to only download a layer's contents once.
+//
+// A localFetchArena must not be copied after first use,
+// and it should be initialized via newLocalFetchArena.
+type localFetchArena struct {
+	// root is the directory to which layers are downloaded.
+	root string
+	// ko is used to ensure each layer is only downloaded once.
+	ko sync.KeyedOnce[string]
+
+	// mu protects rc.
+	mu gosync.Mutex
+	// rc is a map of digest to refcount.
+	rc map[string]int
+}
+
+// newLocalFetchArena initializes a new localFetchArena.
+func newLocalFetchArena(root string) *localFetchArena {
+	return &localFetchArena{
+		root: root,
+		rc:   make(map[string]int),
+	}
+}
+
+// Realizer returns an indexer.Realizer.
+func (f *localFetchArena) Realizer(_ context.Context) indexer.Realizer {
+	return &localFetcher{
+		f: f,
+	}
+}
+
+// Get downloads the image's manifest and returns the related claircore.Manifest.
+//
+// Get also downloads each previously unseen layer of the image into the arena's root directory.
+func (f *localFetchArena) Get(ctx context.Context, image string, opts ...Option) (*claircore.Manifest, error) {
+	// Parse the image name before doing anything else,
+	// as there is no reason to do anything if the image is not properly referenced.
+	ref, err := name.ParseReference(image)
+	if err != nil {
+		return nil, err
+	}
+
+	o := makeOptions(opts...)
+	// Fetch the image's manifest from the registry.
+	desc, err := remote.Get(ref, remote.WithContext(ctx), remote.WithAuth(o.auth), remote.WithPlatform(o.platform))
+	if err != nil {
+		return nil, err
+	}
+
+	img, err := desc.Image()
+	if err != nil {
+		return nil, err
+	}
+	d, err := img.Digest()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
This will perform remote calls to the registry to get the data, is that correct?

In other words, is this where we are pulling the layer or just the image metadata/manifest?